### ISSUE TITLE
[feat]: PostHTML Tree building API

### ISSUE THREAD

**Issue by awinogradov** [posthtml/posthtml#162]
We should make base API for working with PostHTML Tree. It's needed for bigger abstraction with Tree format.

**Comment by awinogradov** [comment]
Simple example https://github.com/theprotein/protein-tree

**Comment by tunnckoCore** [comment]
Firstly we should think for better plugin API. Everything other is easy. Second most important for v1 is better and stable AST. You can't release v1 with that as, it's suicide. That's my 2c.

**Comment by awinogradov** [comment]
@tunnckoCore don't worry! You are absolutely right. We can release this in minor.

**Comment by voischev** [comment]
Profit?

**Comment by awinogradov** [comment]
- Changing AST without pain. Fixed API for plugins and core. Everybody can write custom AST and use it with any plugins or write custom plugins for extended AST. 
- Incapsulate PostHTML Tree structure 
- Minimize core features
- Building PostHTML Tree in other tools and modules

**Comment by awinogradov** [comment]
You can some example of this way in postcss-parser https://github.com/postcss/postcss/blob/master/lib/parser.es6. PostCSS AST has API for build self and now integrated to postcss-core. But you can use it for building custom parsers. I think we can do it better. Small concept example you can see in Protein Tree module https://github.com/theprotein/protein-tree

**Comment by voischev** [comment]
But i think that a distant future.  Before we need stable API in core module ;)

**Comment by awinogradov** [comment]
It's phrase about nothing. What do you mean when say stable API? Any issues?

**Comment by voischev** [comment]
we need more time and cases to separate api to module. You can not cite a single query to whom it would be necessary to. So... that a distant future

**Comment by awinogradov** [comment]
It is the same. You can wait. I can try. You can read again benefits https://github.com/posthtml/posthtml/issues/162#issuecomment-239615276. Then add some troubles with an immutable tree. Then look at https://github.com/posthtml/posthtml/issues/101 ;) Then remember how we extend tree by methods. Now it's ugly way in fact. Then) Answer, please) How we can solve that problem?) This issue about this. Flexible AST with part of basic methods can help to solve many problems in PostHTML. Ideas about waiting are ideas about delaying work only.

**Comment by voischev** [comment]
"Flexible AST" — does not mean that it should be a separate module ;)
Tasks are solved in #101 just a good API in PostHTML. ;)

**Comment by awinogradov** [comment]
Separate module it's not about npm-package. It's a module in general. This module can be in PostHTML, no problem. But it needs to be changeable. You say: we need to wait, we need to make good API and etc. Please, add more concrete things to your answers. Now it looks like you have no plans and you are avoiding of work. Explain your answers by issues and concrete actions to solve problems. It'll be the better way to discuss.

**Comment by voischev** [comment]
Separate also does not seem like a good plan. 

>changeable 

for what? Changeable for changeable?

> not about npm-package

ok! You gave the example protein-tree ;) is a module.

need to wait... to add minimum `tree.render` method and to watch again about this issue

**Comment by awinogradov** [comment]
@voischev you don't want to read benefits? Read them again. I don't see any arguments from you why we don't need to do this proposal. I wrote more than six benefits. It doesn't look like productive discussion. `protein-tree` is an example of realization not of structure or form of it. I don't need your approve to propose this changes. We have more than 2 collaborators. If other guys want this changes we'll include it in the next versions.

**Comment by voischev** [comment]
Concept idea for independent parsers and independent ast in PostHTML

Example for parse5
```js
posthtml([
    'parse5-format-ast-plugin1', 
    'parse5-format-ast-plugin2',
    … 
    'parse5-format-ast-to-posthtml-format-ast', // !!!
    'old-plugin',
    'plugin-for-xjst'
].map(require))
.process(html, { parser: parse5 })
.then(resilt => console.log);
```
if we `content` name node set in `options`
https://github.com/posthtml/posthtml/blob/master/lib/api.js#L71-L72

we can `match` and `walk` api in any AST format.

match api like a parse5
```js
tree.match({ tagName: 'h1', '__location': { line: 5 }}, node => console.log);
```
looks like :)

**Comment by jescalan** [comment]
It seems like it would be difficult for plugins to be able to also process any of these types of trees. How would the plugin know which type of tree it was dealing with? Would every plugin need to include logic to handle many different types of trees? I feel like it's probably a better idea to pick one AST format and go with it.

**Comment by voischev** [comment]
base format PostHTMLTree)

**Comment by jescalan** [comment]
So how would you convert a tree that contains more information than the base posthtml tree format into that format without losing the information?

**Comment by voischev** [comment]
You know about not completeness of PostHTMLTree... in this case need to change the format AST in a major release PostHTML

**Comment by voischev** [comment]
This approach will work with the new tree format, to:)
```js
posthtml([
    'new-format-ast-plugin1', 
    'new-format-ast-plugin2',
    … 
    'new-format-ast-to-posthtml-ast-old-format', // !!!
    'old-plugins'
].map(require))
.process(html, { parser: posthtmlParser2 })
.then(resilt => console.log);
```

**Comment by jescalan** [comment]
Fair enough, but that's a very difficult thing to deal with for users and plugin authors. Figuring out what works with the old/new tree will be challenging. Also what happens if another update is needed to change the tree again? Will you have three different possible formats that need to be converted?

It might be easier to just release a new version with the new tree, and tell people that they can upgrade whenever they are ready. This is really what versioning is for, when it comes down to it. Plugins can use [`peerDependencies`](https://nodejs.org/en/blog/npm/peer-dependencies/) to indicate what version of posthtml they work with natively, so it's clear to users and impossible to make a mistake with compatibility.

**Comment by michael-ciniawsky** [comment]
```js
// Current {Array.<Object|String>}
const tree = [
  '\n',
  { 
    tag: 'name',
    attrs: {
      name: {}
    },
    content: []
  }, 
  '\n'
]

// Maybe {Array.<Object>}
const tree = [
  { content: '\n'},
  { 
    tag: 'name',
    attrs: {
      name: {}
    },
    content: []
  },
  { content: '\n' }
]


```
-Custom Parser-|-------------------- Core ---------------------|-Custom Render--
sugarml ------>| PostHTMLTree -->            ->PostHTMLTree -->| ph-generator 
(Source -> AST)|                |->Plugins->|                  | (Code Gen)
ph-pug ------->| PostHTMLParser->            ->PostHTMLRender->| ph-(js|jsx)
(Source -> HTML)                                                 (Stringifier)
```

```
div(class={test.class})
  h1 {test.text}
```

```js
const pug = require('posthtml-pug')
const jsx = require('posthtml-jsx')

const options = { type: 'es2015', name: 'Test', props: ['test'], export: true }

posthtml()
  .process(html, { parser: pug(), render: jsx(options) })
  .then((result) => result.html)
```

```js
import { Component } from 'react'

class Test extends Component {
  render (test) {
    return (
      <div className={test.class}>
       <h1>{test.text}</h1>
      </div>
    )
  }
}

export default Test
```

**Comment by voischev** [comment]
```js
function convertStringNodeToObject(node) {
    if (typeof node === 'string') return { tag: false, content: node };
    return node;
};

module.export = function pluginMaybeASTFormat = function(tree) {
    tree.walk(convertStringNodeToObject);
};
```

or `posthtml-pug` parser can returns text nodes in object

**Comment by tunnckoCore** [comment]
I started working on something and in next days will show you how the things could be built great. With awesome plugin system and logic everything goes on right way - sync/async plugins, extending the core api, extending/adding AST API, not loading twice same plugin, reporting system and etc.

Called @postcore. Thought for it these days and I came to that in future it can be possible to pass multiple parsers on same input and use different plugins for each project - @postjson, @posthtml and @postcss.

**Comment by michael-ciniawsky** [comment]
@voischev 👍 yep, this adapter could help for eventual future migration, at least over some time until the majority as made the whatever upgrades :)

```js
export default (options = {}) => {
  return function postHTMLPlugin (tree, { walk,  $ }) {
    // Tree API
    walk(tree, ['matcher1', 'matcher2'], node => {
      // some awesome expressive/fluent Node API :)
      $(node)
        .tag('div', (tag) => tag = 'section')
        .attrs('class', (class) => transfrom(class))
        .content({ type: 'tag' }, (ctx) => transform(ctx))
        .extended('whatever', (what) => transform(what))
  }) 
}
```

@tunnckoCore maybe postcore is a little too much :), in my opinion at least the major web technologies (HTML(PostHTML), CSS(PostCSS), JS (Babel), JSON(PostJSON)) should have there clear own tools as modular, flexible and 'low-level' (AST -> transform plugins) as it can be, but go for it, and don't forget to help out here ;). If there is one core using the pattern (Parser -> AST -> AST Plugins -> Output) folks will get even more confused with the matters. There is not one single day at gitter/postcss where not someone comes up with e.g 'I have installed postcss-scss, but it's not compiling my sass files :( etc ..', if they have to use a single core and map the possibilties/ possible setups/configs in their heads that would be overkill :D

**Comment by tunnckoCore** [comment]
Hm. Your example looks like a good idea, using something like `cheerio`.

> should have there clear own implementations as modular

What so own? Methods and functionality is the same (or should be), diffs are the ASTs.

> There is not one single day at gitter/postcss where not someone comes up with

because implementations sucks, not because some API or AST things.

**Comment by michael-ciniawsky** [comment]
> Hm. Your example looks like a good idea, using something like cheerio

yep, there was/is a lib clalled postXML using cheerio to do whats posthtml attempts to do, but specific functional util lib for working with the AST is superior imho

> What so own? Methods and functionality is the same (or should be), diffs are the ASTs.

technically 100% right, but get the different established communties the shift to a unified module will be hard and a lot of chore work. Besides of that a core lib implementing the interface and a derivate for each technologie thats a nice idea in general. The interfaces are mostly similiar anyways, which is good thing :). But in terms of ecosystems  let's take e.g simply searching for plugins with clear prefixes ```posthtml-*```, ```postcss-*```, ```postjson-*``` instead of ```postcore-html-[parser|plugin|render]-[name], postcore-css-[parser|plugin|render]-[name]```. Core derviates could/should ship with a default parser and render aswell ;).

```js
postcore(plugins)
  .use({ parser: html(), render: html() } || (tree) => tree)
  .parse(file.[ext])
  .render(tree)
  .process(file.[ext])
    .then((result) => result) // async
    .result // sync
````

```js
class PosHTML extends PostCore {}
class PostCSS extends PostCore {}
.....
```

> Yea, absolutely.

👍

**Comment by voischev** [comment]
@tunnckoCore look https://github.com/jstransformers/

**Comment by voischev** [comment]
postcore = gulp?)))

**Comment by tunnckoCore** [comment]
@michael-ciniawsky it's not needed to change original prefixes and etc. PostCore not need fame, it wil be just engine behind the scenes such as JSTransformer and `standard-engine`, or wooorm's retext, rehype and etc - don't remember what was the name behind them - `unist` i suppose? nvm, but his plugins system isn't the best too.

@voischev what to look? i'm there from the beginning.

Can't see what similar you see between @postcore -> @jstransformers, and @postcore -> @gulpjs

**Comment by michael-ciniawsky** [comment]
it's not about 'fame' per se. But if there is not a unification with the different user bases, then it's technically unified but nevertheless introduces 'yet a another' module to to what has to be done :).

> postcore = gulp?)))

yep, in some way it gets a slightly smell of a build tool, but 
anyways for now the future AST format/Tree API nerds and nerdines...:D

**Comment by wooorm** [comment]
@tunnckoCore The nodes are called [Unist](https://github.com/wooorm/unist) (the HTML namespace [HAST](https://github.com/wooorm/hast)), the processors are [unified](https://github.com/wooorm/unified)!

**Comment by tunnckoCore** [comment]
@wooorm yea yea, checked it after I commented, thanks. Hm, didn't look it recently and it seems better than last time. :)

**Comment by michael-ciniawsky** [comment]
@wooorm 👋 I have a question and wanted to ask for a while now 😛  would it be possible to/of use to consider [vfile](https://github.com/wooorm/vfile) for posthtml ? and what should/could a new PostHTML AST look like in your opinion?

**Comment by tunnckoCore** [comment]
@michael-ciniawsky yea, it's possible. the whole thing around @wooorm with that things is great work these years. Was totally forgot that i reviewed everything he did on such processing.

**Comment by wooorm** [comment]
Cool! Appreciate it!

I don’t check all issues here so I’m not entirely sure about the current state of PostHTML things though.

Yes, `vfile` is definitely usable. `vfile` is very much like Vinyl, especially if [2.0.0](https://github.com/wooorm/vfile/pull/10) is released (somewhere this/next week?) Difference being the streams are dropped, a more minimal API, and pretty powerful metadata support.

Re AST: HAST is currently using `properties` (`className` instead of `class`), and property values (`className` is `Array.<string>`). It’s cool that it’s more like JS, but it needs a big library of how props are capitalised, and which props are booleans / numbers / lists ([`property-information`](https://www.npmjs.com/package/property-information)).

I personally suggest either using HAST, or creating a similar namespace which uses `element.attributes` instead of `element.properties`, and has `Object.<string>` instead of `Object.<(string|boolean|number)?|Array.<(string|boolean|number)?>>`

**Comment by tunnckoCore** [comment]
@michael-ciniawsky Yea, definitely implementing HAST would be the best.
@wooorm, the only thing that I don't like about that plugin system is the signature - it should be (and it will be better) to be concrete (as you write it in code comments). Some like that 

```js
module.exports = function foobar (opts) {
  return function plugin (app) {
    // `this` context is same as `app`
    app.options = extend({}, app.options, opts)
    app.addQux = function () {}

    // called by the `.run(foo, bar, baz)` method
    return function qux (ctx, tree, file, etc) {
      ctx.options = extend({}, ctx.options, app.options)
      // stuff
    }
  }
}

// so then
app.use(foobar({ options: 'if needed' }))
app.addQux()
```

This signature is awesome for couple of reasons. Enables you to extend core `app`, so.. plugins for the app, no only for processing text using ctx/file/etc. Hance, look the @node-base you'll love its architecture - it is event emitter with few builtin useful methods, and built in registering system (meaning that plugin will be loaded only once, but you also can force loading it multiple times).
Built with @node-base are Verb, Generate, Update and Assemble.

**Comment by wooorm** [comment]
I appreciate the comments but it’s quite off topic, let’s talk about the AST here :)

---

That can currently be written something like:

```js
module.exports = function foobar (app, opts) {
  // `this` context is same as `app`
  // ...

  // called by the `.run()` method
  return function qux () {
    // ...
  }
}

// so then
app.use(foobar, { options: 'if needed' }).run(/*...*/)
// ...
```

I have no idea what this code is though. It contains lots of stuff I’m
unfamiliar with. **unified** is for processing files. Not for creating web servers.

> This signature is awesome for couple of reasons. Enables you to extend core app, so.. plugins for the app, no only for processing text using ctx/file/etc.

Both `this` and the first arg are the processor.

> Hance, look the @node-base you'll love its architecture - it is event emitter with few builtin useful methods...

unified is as well, haven’t found a use for events though.

> ...and built in registering system (meaning that plugin will be loaded only once, but you also can force loading it multiple times).

I haven’t needed that logic.

>  Built with @node-base are Verb, Generate, Update and Assemble. Not promoting and forcing anything, just they need some word and share :)

:+1:

**Comment by michael-ciniawsky** [comment]
@wooorm 

> I appreciate the comments but it’s quite off topic, let’s talk about the AST here :) 

👍  I will comment on Vinly 2.0.0 related to a few naming things in the direction sindre suggested would be awesome. i have also questions in regard to alex :), but haven't found the time yet to call out for that

- Can vfile contain/update metadata about location info changes (diffs) made by plugins for sourcemappings?

```txt
AST -> Plugin 1 - Plugin2 -> Plugin3 -> HTML
vfile.?map -> update(vfile.?map) - update(vfile.?map)- update(vfile.?map) -> vfile.?map  
```

```js
// HAST
{
  "type": "element",
  "tagName": "a",
  "properties": {
    "href": "http://alpha.com",
    "id": "bravo",
    "className": ["bravo"],
    "download": true
  },
  "children": []
}

// PHAST :)
[
{
  "type": "element",
  "tag": "a",
  "attrs": {
    "href": "http://alpha.com",
    "id": "bravo",
    "class": ["bravo"],
    "download": true
  },
  "content": []
},
{ 
  type: 'text',
  content: 'Allo Allo! :)"
}
]
```

- Where should location info find its place ?-> vfile.?location?
- What about whitespace/indentation for e.g linting and including partials and update them with the right indent?

**Comment by wooorm** [comment]
>  i have also questions in regard to alex 

Either raise an issue or chat/mail me!

> Can vfile contain/update metadata about location info changes (diffs) made by plugins for sourcemappings?

Haven’t checked that, actually. Is there actually an example of HTML source maps?

> Where should location info find its place ? -> vfile.?location?

Locations are in each node, at `node.position`. Read more about that [here](https://github.com/wooorm/unist#location).

> What about whitespace/indentation for e.g linting and including partials and update them with the right indent?

In HTML / HAST, they’re [text](https://github.com/wooorm/hast#text) nodes. I have a half working minifier for rehype locally, stripping white space smartly, and from what I checked the inverse (prettifying) is also possible!

> ```
{
  "type": "element",
  "tag": "a",
  "attrs": {
    "href": "http://alpha.com",
    "id": "bravo",
    "class": ["bravo"],
    "download": true
  },
  "content": []
},
{ 
  type: 'text',
  content: "Allo Allo! :)"
}
```

I suggest:

```diff
{
  "type": "element",
  "tag": "a",
  "attrs": {
    "href": "http://alpha.com",
    "id": "bravo",
-   "class": ["bravo"],
+   "class": "bravo", // otherwise you need a list of values which accept a space-separated, or
+                     // comma-separated list
-   "download": true
+   "download": "download" // or `""`, same as above: you’ll need a list of attribute names.
  },
- "content": []
+ "children": []
},
{ 
  type: 'text',
-  content: "Allo Allo! :)"
+  value: "Allo Allo! :)"
}
```

I suggest ~~content~~ > children and ~~content~~ > value because it allows integration with all these [utilities](https://github.com/wooorm/unist#list-of-utilities).

**Comment by michael-ciniawsky** [comment]
> Either raise an issue or chat/mail me! 

yep 👍 , it there a gitter fro alex/franc? :)

linter-alex was my nemesis at first, a everything is dicussable dev world. 

```would be can weaking meaning....``` :D :D i need to hang this over my bed

> Haven’t checked that, actually. Is there actually an example of HTML source maps?

nope, to be honest i'm even not quite sure if they make sense for html at all, it came up over time so i started 'researching' about them but haven't found anything valueable. But anyways throwing it in. Your doubts clearly strengthened mine for example :).

```js
[{
  "parent": 'body' // ? or root: 'body'
  "type": "element",
  "space": '  '
  "tag": "a",
  "attrs": {
    "href": { value: "http://alpha.com", position: {}} // ? position, i mean node.attrs[name] = 'value' is nice to work with :) 
    "id": { value: "bravo", position: {}}
     "class": { value:"bravo bravo", position: {}} // simply split if multiple classes are on the node 
     "download": { value: "", position: {}}
  },
  "children": []
   "position": { line: '' col: '', offset: '' }
},
{ 
  "type": 'text',
  "space": '  ', // ?
  "value": "Allo Allo! :)"
  "position": { line: '' col: '', offset: '' }
}]
[file] : new Vfile('path/to/file.html') 
// => vfile { cwd: process.cwd(), path: path/to/file.html, dir: path/to, name: file ext: 'html' } ?
[walk]
[match]
```

> In HTML / HAST, they’re text nodes. I have a half working minifier for rehype locally, stripping white space smartly, and from what I checked the inverse (prettifying) is also possible!

Are there any infos/insights avaiable for the average joe at the moment :D ?

**Comment by wooorm** [comment]
> yep 👍 , it there a gitter for alex/franc? :)

Nope, haven’t found the need yet. Not much chatter related to it. http://gitter.im/wooorm is ok though.

> linter-alex was my nemesis at first, a everything is dicussable dev world.

👌 

> nope, to be honest i'm even not quite sure if they make sense for html at all, it came up over time so i started 'researching' about them but haven't found anything valueable. But anyways throwing it in. Your doubts clearly strengthened mine for example :).

Someone asked for that in remark a while back; I didn’t find any browser that supports it, or library the exposes them for HTML. So I don’t see why to investigate further (pretty sure it’s possible though).

> `  "parent": 'body' // ? or root: 'body'`

Why need that? Keeping track of that just makes it harder to move tags around (like replacing `<i>` with `<em>`, it needs extra bookkeeping), and that info is already available if you’re walking the tree!

> `  "space": '  ' // ?`

Why not have a separate node for that? Note that space is pretty important in HTML (like between two nodes which are rendered inline, possible even inlined through CSS).

> ```
  "attrs": {
    "href": { value: "http://alpha.com", position: {}} // ? position, i mean node.attrs[name] = 'value' is nice to work with :) 
    "id": { value: "bravo", position: {}}
    "class": { value: "bravo bravo", position: {}} // simply split if multiple classes are on the node.attrs.class.value 
    "download": { value: "", position: {}}
  },
```

Attributes as objects: That’s pretty verbose, and it doesn’t work well with [“generated”](https://github.com/wooorm/unist#location) nodes, or those coming from e.g. React / Pug / &c. Here’s how I [did it in rehype](https://github.com/wooorm/rehype/tree/master/packages/rehype-parse#optionsverbose)

> Are there any infos/insights available for the average joe at the moment :D ?

About rehype? rehype-minify?

**Comment by michael-ciniawsky** [comment]
> Nope, haven’t found the need yet. Not much chatter related to it. http://gitter.im/wooorm is ok though.

:+1: give some time to concrete and recap  on that, it's basically about browser/client side usage for comments (general language aid, code of conduct, discussion rules) etc...

> Someone asked for that in remark a while back; I didn’t find any browser that supports it, or library the exposes them for HTML. So I don’t see why to investigate further (pretty sure it’s possible though).

Same here, there was a issue asking for support an following discussion, resulting in the same conclusion it's not a top priority but if possible keep on the list.

> Why need that? Keeping track of that just makes it harder to move tags around (like replacing ```<i>``` with ```<em>```, it needs extra bookkeeping), and that info is already available if you’re walking the tree!

Yep true, let's ditch that :)

> Why not have a separate node for that? Note that space is pretty important in HTML (like between two nodes which are rendered inline, possible even inlined through CSS).

> Here’s how I did it in rehype

Both 👌 :)

> Attributes as objects: That’s pretty verbose, and it doesn’t work well with “generated” nodes, or those coming from e.g. React / Pug / &c. Here’s how I did it in rehype

Yep, personally also not happy with objects as values

> About rehype? rehype-minify?

How the prettify and minification works in detail 

```js
[
  {
     type: 'space',
     value: '  ', // \n\t etc aswell? 
     position: [{ line: '', col: '', offset: '' }, {line: '', col: '', offset: '' }]
  }, 
  { 
     type: 'element'
     tag: 'name'
     attrs: {
           name: 'value'
     }
     value: 'Allo Allo :)',  // how is text/content represented ? :)
     children: [],  
     position: {
         tag: [{ line: '', col: '', offset: '' }, {line: '', col: '', offset: '' }]
         attrs: {
           name: [
             {line: '', col: '', offset: '' } // start [0]
             {line: '', col: '', offset: '' } // end   [1]
           ]
         },
         value: [{ line: '', col: '', offset: '' }, {line: '', col: '', offset: '' }]
      }
   },
   {
     type: 'space',
     value: '    ',
     position: [{ line: '', col: '', offset: '' }, {line: '', col: '', offset: '' }]
   }, 
   {
     type: 'text',
     value: 'Allo Allo :)',
     position: [{ line: '', col: '', offset: '' }, {line: '', col: '', offset: '' }]
   }
]
```

**Comment by wooorm** [comment]
> 👍 give some time to concrete and recap on that, it's basically about browser/client side usage for comments (general language aid, code of conduct, discussion rules) etc...

Cool! This lists a few projects which do similar things: https://github.com/wooorm/remark/blob/master/doc/products.md#other
Also: alex is just remark/retext and some plug-ins, and there’s Gitter rooms for both!

> How the prettify and minification works in detail

Haha, well in detail: it isn’t finished yet. I can see if I can upload the unfinished repo soon, and give you a preview.

In short, take this document:

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Page  Title</title>
  </head>
  <body>
    <h1>Hello  World!</h1>
  </body>
</html>
```

That yields the following tree:

```js
{ type: 'root',
  children: 
   [ { type: 'doctype',
       name: 'html',
       public: null,
       system: null,
       position: 
        { start: { line: 1, column: 1, offset: 0 },
          end: { line: 1, column: 16, offset: 15 } } },
     { type: 'element',
       tagName: 'html',
       properties: { lang: 'en' },
       children: 
        [ { type: 'element',
            tagName: 'head',
            properties: {},
            children: 
             [ { type: 'text',
                 value: '\n    ',
                 position: 
                  { start: { line: 3, column: 9, offset: 41 },
                    end: { line: 4, column: 5, offset: 46 } } },
               { type: 'element',
                 tagName: 'meta',
                 properties: { charSet: 'utf-8' },
                 children: [],
                 position: 
                  { start: { line: 4, column: 5, offset: 46 },
                    end: { line: 4, column: 27, offset: 68 } } },
               { type: 'text',
                 value: '\n    ',
                 position: 
                  { start: { line: 4, column: 27, offset: 68 },
                    end: { line: 5, column: 5, offset: 73 } } },
               { type: 'element',
                 tagName: 'title',
                 properties: {},
                 children: 
                  [ { type: 'text',
                      value: 'Page  Title',
                      position: 
                       { start: { line: 5, column: 12, offset: 80 },
                         end: { line: 5, column: 23, offset: 91 } } } ],
                 position: 
                  { start: { line: 5, column: 5, offset: 73 },
                    end: { line: 5, column: 31, offset: 99 } } },
               { type: 'text',
                 value: '\n  ',
                 position: 
                  { start: { line: 5, column: 31, offset: 99 },
                    end: { line: 6, column: 3, offset: 102 } } } ],
            position: 
             { start: { line: 3, column: 3, offset: 35 },
               end: { line: 6, column: 10, offset: 109 } } },
          { type: 'text',
            value: '\n  ',
            position: 
             { start: { line: 6, column: 10, offset: 109 },
               end: { line: 7, column: 3, offset: 112 } } },
          { type: 'element',
            tagName: 'body',
            properties: {},
            children: 
             [ { type: 'text',
                 value: '\n    ',
                 position: 
                  { start: { line: 7, column: 9, offset: 118 },
                    end: { line: 8, column: 5, offset: 123 } } },
               { type: 'element',
                 tagName: 'h1',
                 properties: {},
                 children: 
                  [ { type: 'text',
                      value: 'Hello  World!',
                      position: 
                       { start: { line: 8, column: 9, offset: 127 },
                         end: { line: 8, column: 22, offset: 140 } } } ],
                 position: 
                  { start: { line: 8, column: 5, offset: 123 },
                    end: { line: 8, column: 27, offset: 145 } } },
               { type: 'text',
                 value: '\n  \n\n',
                 position: 
                  { start: { line: 8, column: 27, offset: 145 },
                    end: { line: 11, column: 1, offset: 164 } } } ],
            position: 
             { start: { line: 7, column: 3, offset: 112 },
               end: { line: 9, column: 10, offset: 155 } } } ],
       position: 
        { start: { line: 2, column: 1, offset: 16 },
          end: { line: 10, column: 8, offset: 163 } } } ],
  data: { quirksMode: false },
  position: 
   { start: { line: 1, column: 1, offset: 0 },
     end: { line: 11, column: 1, offset: 164 } } }
```

First, that tree is transformed to remove useless things: `utf-8` > `utf8`, and multiple spaces `page  title` to one `page title`. Then, there’s quite a lot of logic needed to remove other white space, like the one in `head` (it’ll never show), or the ones before/after/between block levels nodes, &c.
Finally, there the compilation phase, which smartly omits quotes, tags, and more.

```html
<!DOCTYPE html><html lang=en><meta charset=utf8><title>Page Title</title><h1>Hello World!</h1>
```

But it isn’t done yet. More will come! 😏

**Comment by michael-ciniawsky** [comment]
> Cool! This lists a few projects which do similar things: https://github.com/wooorm/remark/blob/master/doc/products.md#other
Also: alex is just remark/retext and some plug-ins, and there’s Gitter rooms for both!

@ wooorm :+1: yep thats all the recaps i must make, i don't have a sophistiated knowledge about it . I head over to to who of them when ready :)

@posthtml/collaborators 

## Nodes

###  Doctype

```js
{
   type: 'doctype'
   version: 'html',
   pid: 'null',
   uid: 'null',
   position: [ 
    {line: '', col: '', [offset: '']}, // offset ? true : false
    {line: '', col: '', [offset: '']}
  ]
}
```

### Tag 

```js
{
   type: 'tag'
   tag: 'name',
   attrs: { name: {} }
   content: [] // children ?
   position: { 
     tag: [{}, {}], // { start: {}, end: {} } ?
     attrs: { name: [{}, {}] }
   }
}
```

### EOL

```js
{
   type: 'eol' // 'merge into' node.type = 'text' ?
   content: '\n',
   position: [{}, {}]
}
```

### Space

```js
{
   type: 'space' // 'merge into' node.type = 'text' ?
   content: '  ',
   position: [{}, {}]
}
```

### Text

```js
{
   type: 'text'
   content: 'Allo Allo :)',
   position: [{}, {}]
}
```

### Comment

```js
{
   type: 'comment',
   content: "I'm a comment",
   position: [{}, {}]
}
```

**Comment by michael-ciniawsky** [comment]
@posthtml/collaborators ping ping ping :). We should at least concretize a possible future format without commitment to it for now.

**Comment by jescalan** [comment]
You guys already generally know my thoughts on it, you can see the full docs for the AST format I worked on [here](https://github.com/reshape/reshape#reshape-ast), which is fully functional and already working well. But some quick feedback on what was proposed above:

- For the doctype, it's `pid` and `sid`, the latter is "system id"
- It's not possible for a single tag to have multiple positions, so no need to make this an array, also don't understand the purpose of `offset`. If you mean `start` and `end` in the 2+ positions, I would question what you need the end for exactly. If you have a valid use-case for end, keep it, but make it more clear by making it an object with `start` and `end` keys rather than an array.
- Why do you need different nodes for `text`, `eol`, and `space`? They are all text nodes, that could be easily categorized by plugin authors using a short regex. Also, how will you parse this correctly? What if there is a text node that contains newlines and/or spaces? Does it get broken up into a separate node for every space or newline it contains? This will make it very difficult for plugin authors to handle text contents.

**Comment by tunnckoCore** [comment]
:+1: with @jescalan

`doctype` and `tag` nodes seems strange. Not make sense position to be array, this one `{ tag: [{}, {}], // { start: {},end: {} } ?attrs: { name: [{}, {}] }}` also.

Actually I was totally with @jescalan for his PR.

But don't like that he created 3rd html post processor - reshape, posthtml and rehype.

**Comment by jescalan** [comment]
@tunnckoCore Sorry, I tried not to! In the end, I didn't really have much of an option. Still am happy to contribute here where needed, and when reshape gets a public release, it will be production-ready and heavily marketed.

**Comment by michael-ciniawsky** [comment]
> For the doctype, it's pid and sid, the latter is "system id"

yep, typo... corrected

The other things are mostly the mention them with maxium verbosity, and simplify and ditch unnecessity during discussion, i'm mostly fine and like reshape AST format, ```node.name``, {Array} as attrs.name ```attrs: { id: [{}] // ? or is this only for attr.class in case there a mutliple classes?``` and maybe the position/location info at one place, but the later is not that important.

Others need/should have the opportunity to share their 2 cents here, i will ping here every day, you have been warned :)

> Sorry, I tried not to! When reshape gets a public release, it will be production-ready and heavily marketed.

mhmmm.... here is your fish  🎣  👾   😛

**Comment by michael-ciniawsky** [comment]
@tunnckoCore 

```doctype``` yep, typo is corrected
```tag``` agreed, either ```{start: {}, end: {}}```or directly as ```node.line node.col``` without ```end: {}``` analog to reshape AST.
```?attrs: { name: [{}, {}] }}``` typo aswell should have been ```attrs: { name: {} }``` of course

But don't like that he created 3rd html post processor - reshape, posthtml and rehype.

Yep 👍 , completely unnecesscary, should be reconcilated when possible imho

**Comment by voischev** [comment]
@michael-ciniawsky my ideas about format in this gist https://gist.github.com/voischev/767c2822b6fe12f1ac608aba2b1cc888  (wip)

**Comment by michael-ciniawsky** [comment]
Any progress here ? :)

@voischev I have `node.type`, `node.position` in a backwards compatible manner 90% ready for posthtml-parser (with htmlparser2), has a beginning :) 

```js
"\n  ",
      {
        "type": "tag",
        "tag": "div",
        "attrs": {},
        "content": [
          "\n    ",
          {
            "type": "tag",
            "tag": "span",
            "attrs": { "id": "test" },
            "content": [ "Text1" ],
            "position": {
              "start": { "line": 8, "col": 6 },
              "attrs": { "id": { "line": 8,  "col": 17 }, ...},
              "end": { "line": 8, "col": 24 }
            }
          },
          "\n    ",
        ],
        "position": { "start": { "line": 7, "col": 5 },  "end": { "line": 13, "col": 6 } }
      },
"\n  ",
```

**Comment by voischev** [comment]
not actual
